{% macro make_db_link(db) %}
{% if db.name == 'resfinder' %}
    {% set base_url="bitbucket.org/genomicepidemiology/resfinder_db/commits" %}
{% elif db.name == 'virulencefinder' %}
    {% set base_url="bitbucket.org/genomicepidemiology/virulencefinder_db/commits" %}
{% elif db.name == 'pointfinder' %}
    {% set base_url="bitbucket.org/genomicepidemiology/pointfinder_db/commits" %}
{% endif %}
<a class="br-card-link" href="https://{{base_url}}/{{db.version}}" target="_blank">{{ db.version[:6] }}</a>
{% endmacro %}


{% macro sample_header(sample, class_names="br-header") %}
<div class="p-2 {{class_names}}">
    <div class="card">
        <div class="card-body">
            <div class="d-flex">
                <h4 id="sample-id-heading" class="card-title">{{sample.sample_id}}</h4>
                {% for tag in sample.tags %}
                    {% if tag.type == "virulence" %}
                        {% set heading="virulence-heading" %}
                    {% else %}
                        {% set heading="phenotype-heading" %}
                    {% endif %}
                    <span 
                        class="tag badge ms-2 rounded-pill bg-{{tag.severity}}"
                        data-bs-toggle="tooltip" data-bs-placement="top"
                        {% if tag.description %}data-bs-title="{{ tag.description }}"{% endif %}
                    >
                        {{tag.label | upper }}
                    </span>
                {% endfor %}
            </div>
            <div class="row gx-5">
                <div class="col-md-auto">
                    Upload date: {{sample.created_at | strftime}}
                    <br>
                    <p class="card-text">
                        {% if sample.qc_status.status == "passed" %}
                            QC: <strong class="text-success">{{ sample.qc_status.status | capitalize }}</strong>
                        {% elif sample.qc_status.status == "failed" %}
                            QC: <strong class="text-danger">{{ sample.qc_status.status | capitalize }}</strong>; {{ sample.qc_status.action }}, {{ sample.qc_status.comment }}
                        {% else %}
                            QC: <strong>{{ sample.qc_status.status | capitalize }}</strong>
                        {% endif %}
                    </p>
                </div>
            </div>
        </div>
    </div>
</div>
{% endmacro %}


{% macro accession_link(accession, name=None) %}
    <a class="br-card-link" target="_blank" href="https://www.ncbi.nlm.nih.gov/protein/{{accession}}">
        {% if name == None %}
            {{ accession }}
        {% else %}
            {{ name }}
        {% endif %}
    </a>
{% endmacro %}

{% macro resistance_summary_row(gene)%}
    <td>{{ accession_link(gene.accession) }}</td>
    <td>{{ gene.depth | fmt_null_values }}</td>
    <td>{{ gene.identity | fmt_number }}%</td>
    <td>{{ gene.coverage | fmt_number }}%</td>
    <td class="fst-italic fw-lighter">{{ gene.software }}</td>
{% endmacro %}

{% macro resistance_summary_card(resistance, summary=None) %}
<div class="card">
    <div class="card-header">AMR profile summary</div>
    <div class="card-body">
        <div class="table-responsive">
            <table class="table table-hover">
                <thead>
                    {% for antib_class in summary %}
                        <th>{{ antib_class | capitalize }}</th>
                    {% endfor %}
                </thead>
                <tbody>
                    {% for antibs in zip_longest(*summary.values()) %}
                        <tr>
                            {% for antib in antibs %}
                            <td>
                                {% if antib %}
                                    {{ antib.gene_symbol }}
                                    <span class="badge rounded-pill bg-secondary">
                                        {{antib.software | length}}
                                    </span>
                                {% endif %}
                            </td>
                            {% endfor %}
                        </tr>
                    {% endfor %}
                </tbody>
            </table>
        </div>
    </div>
</div>
{% endmacro %}


{% macro resistance_table_card(resistance, summary=None) %}
<div class="card">
    {% set res=resistance.result%}
    <div class="card-header">{{ resistance.type | camelcase_to_text | capitalize }}</div>
    <div class="card-body">
        <h5 class="card-title">Predicted genes</h5>
        <table class="table">
            <thead>
                <tr>
                    <td scope="col">Gene</td>
                    <td scope="col">Name</td>
                    <td scope="col">Class</td>
                    <td scope="col">Accession</td>
                    <td scope="col">Identity</td>
                    <td scope="col">Coverage</td>
                </tr>
            </thead>
            <tbody>
            {% for gene in res.genes %}
            <tr>
                <td>{{ gene.gene_symbol }}</td>
                <td>{{ gene.sequence_name }}</td>
                <td>
                    {% if gene.res_class == gene.res_subclass %}
                        {{ gene.res_subclass | lower | capitalize }}
                    {% else %}
                        {{ gene.res_class | lower | capitalize }} / {{ gene.res_subclass | lower | capitalize }} 
                    {% endif %}
                </td>
                <td>{{ accession_link(gene.accession) }}</td>
                <td>{{ gene.coverage }}</td>
                <td>{{ gene.identity }}</td>
            </tr>
            {% endfor %}
            </tbody>
        </table>
        <h5 class="card-title">Predicted mutations</h5>
            {% if resistance.mutations | length == 0 %}
                <p class="fw-light fst-italic">No resistance yeilding mutations identified</p>
            {% else %}
                <table class="table">
                    <thead>
                        <tr>
                            <td scope="col">Mutation</td>
                            <td scope="col">Change</td>
                            <td scope="col">Type</td>
                            <td scope="col">Gene</td>
                            <td scope="col">Depth</td>
                        </tr>
                    </thead>
                    <tbody>
                    {% for mut in res.mutations %}
                    <tr>
                        <td>{{ mut.refCodon }}<b>{{ mut.position }}</b>{{ mut.altCodon }}
                            <span class="badge bg-danger" data-bs-toggle="tooltip" data-bs-placement="top" 
                            >
                                R {{ mut.phenotypes | length }}
                            </span>
                        </td>
                        <td>{{ mut.ref_codon | nt_to_aa }}<i class="bi bi-arrow-right mx-2"></i>{{ mut.altCodon | nt_to_aa }}</td>
                        <td>{{ mut.variant_type }}</td>
                        <td>{{ ", ".join(mut.genes) }}</td>
                        <td>{{ mut.depth }}</td>
                    </tr>
                    {% endfor %}
                    </tbody>
                </table>
            {% endif %}
    </div>
</div>
{% endmacro %}


{% macro virulence_card(pred) %}
<div class="card">
    <div class="card-header">{{ pred.software | capitalize }} [ <i>{{pred.type}}</i> ]</div>
    <div class="card-body">
    {% if pred.software == "virulencefinder" %}
        {{ virulencefinder_prediction(pred.result) }}
    {% elif pred.software == "amrfinder" and pred.type == "VIRULENCE" %}
        {{ amrfinder_virulence_prediction(pred.result) }}
    {% endif %}
    </div>
</div>
{% endmacro %}


{% macro amrfinder_virulence_prediction(res, extended=False) %}
<table class="table table-striped">
    <thead>
        <tr>
            <td scope="col">Gene</td>
            <td scope="col">Method</td>
            {% if extended %}
                <td>Contig</td>
                <td>Start/ End</td>
            {% endif  %}
            <td scope="col">Identity</td>
            <td scope="col">Coverage</td>
            <td scope="col">Name</td>
        </tr>
    </thead>
    <tbody class="table-group-divider">
    {% for gene in res.genes %}
    <tr>
        <td>{{ accession_link(gene.accession, name=gene.gene_symbol) }}</td>
        <td>
            {{ gene.method }}
            <i class="bi bi-question-circle"
                data-bs-toggle="tooltip"
                data-bs-placement="top"
                data-bs-html="true"
                data-bs-title="{{ amrfinder_method_desc(method=gene.method) }}"></i>
        </td>
        {% if extended %}
            <td class="tbl-border-left">{{ gene.contig_id }}</td>
            <td class="tbl-border-right">{{ gene.ass_start_pos }}/ {{ gene.ass_end_pos }}</td>
        {% endif  %}
        <td>{{ gene.coverage }}%</td>
        <td>{{ gene.identity }}%</td>
        <td>{{ gene.sequence_name  | capitalize }}</td>
    </tr>
    {% endfor %}
    </tbody>
</table>
{% endmacro %}


{% macro virulencefinder_prediction(res, extended=False) %}
<table class="table table-striped">
    <thead>
        <tr>
            <td scope="col">Gene</td>
            {% if extended %}
                <td>Contig</td>
                <td>Start/ End</td>
            {% endif  %}
            <td scope="col">Identity</td>
            <td scope="col">Coverage</td>
            <td scope="col">Function</td>
        </tr>
    </thead>
    <tbody class="table-group-divider">
    {% for gene in res.genes %}
    <tr>
        <td>{{ accession_link(gene.accession, name=gene.gene_symbol) }}</td>
        {% if extended %}
            <td class="tbl-border-left">{{ gene.contig_id }}</td>
            <td class="tbl-border-right">{{ gene.ass_start_pos }}/ {{ gene.ass_end_pos }}</td>
        {% endif  %}
        <td>{{ gene.coverage }}%</td>
        <td>{{ gene.identity }}%</td>
        <td>{{ gene.sequence_name | capitalize }}</td>
    </tr>
    {% endfor %}
    </tbody>
</table>
{% endmacro %}


{% macro species_prediction_card(predictions, extended=False) %}
<div class="card">
    <div class="card-header">Species prediction</div>
    <div class="card-body">
        <table class="table table-sm">
            <thead>
                <th>Specie</th>
                <th>Fraction of reads</th>
                {% if extended %}
                    <th>Assigned reads</th>
                    <th>Added reads</th>
                {% else %}
                    <th>N reads</th>
                {% endif %}
            </thead>
            <tbody>
                {% for pred in predictions %}
                    {% if extended or pred.fraction_total_reads > 0.001 %}
                        <tr class="{% if predictions | length > 1 and loop.index == 1 %}table-success{% endif %}">
                            <td> <a class="br-card-link fst-italic" 
                                href="https://www.ncbi.nlm.nih.gov/datasets/taxonomy/{{pred.taxonomy_id}}/" 
                                target="_blank">{{pred.scientific_name}}</a></td>
                            <td>{{ (pred.fraction_total_reads * 100) | fmt_number(2) }}%</td>
                            {% if extended %}
                                <td>{{ pred.kraken_assigned_reads | fmt_to_human_readable }}</td>
                                <td>{{ pred.added_reads | fmt_to_human_readable }}</td>
                            {% else %}
                                <td>
                                    {{ (pred.kraken_assigned_reads + pred.added_reads) | fmt_to_human_readable }}
                                </td>
                            {% endif %}
                        </tr>
                    {% endif %}
                {% endfor %}
            </tbody>
        </table>
    </div>
</div>
{% endmacro %}


{% macro amr_prediction_card(pred, extended=False) %}
<div class="card">
    <div class="card-header">{{ pred.software | capitalize }} [ <i>{{pred.type}}</i> ]</div>
    <div class="card-body">
    {% if pred.software == "resfinder" %}
        {{ resfinder_prediction(pred.result, extended=extended) }}
    {% elif pred.software == "amrfinder" and pred.type == "AMR" %}
        {{ amrfinder_amr_prediction(pred.result, extended=extended) }}
    {% elif pred.software == "amrfinder" and pred.type == "STRESS" %}
        {{ amrfinder_stress_prediction(pred.result, extended=extended) }}
    {% elif pred.software == "mykrobe" %}
        {{ mykrobe_amr_prediction(pred.result, extended=extended) }}
    {% elif pred.software == "tbprofiler" %}
        {{ tbprofiler_amr_prediction(pred.result, extended=extended) }}
    {% endif %}
    </div>
</div>
{% endmacro %}


{% macro resfinder_prediction(result, extended=False) %}
<div>
    <h5 class="card-title">Genes</h5>
    <table class="table table-striped">
        <caption>Database: {{result.genes[0].ref_database}}</caption>
        <thead>
            <td scope="col">Gene</td>
            <td scope="col">Accession</td>
            <td scope="col">Identity</td>
            <td scope="col">Coverage</td>
            <td scope="col">Depth</td>
            <td scope="col">Antibiotics</td>
        </thead>
        <tbody class="table-group-divider">
        {% for gene in result.genes %}
            <tr>
                <td>{{ gene.gene_symbol }}</td>
                <td>{{ accession_link(gene.accession) }}</td>
                <td>{{ gene.coverage }}%</td>
                <td>{{ gene.identity }}%</td>
                <td>{{ gene.depth | fmt_number }}</td>
                <td>{{ gene.phenotypes | get_resistance_profile(level="name") }}</td>
            </tr>
        {% endfor %}
        </tbody>
    </table>
    <h5 class="card-title">Variants</h5>
    {% if result.variants | length == 0 %}
        <p class="fw-light fst-italic">No resistance yeilding mutations identified</p>
    {% else %}
        <table class="table table-striped">
            <caption>Database: {{ result.variants[0].ref_database }}</caption>
            <thead>
                <tr>
                    <td scope="col">Variant</td>
                    <td scope="col">Gene</td>
                    <td scope="col">Protein change</td>
                    <td scope="col">Depth</td>
                    <td scope="col">Antibiotics</td>
                </tr>
            </thead>
            <tbody class="table-group-divider">
                {% for variant in result.variants %}
                    <tr>
                        <td>{{ variant.name }}</td>
                        <td>{{ accession_link(variant.close_seq_name, name=variant.gene_symbol) }}</td>
                        <td>{{ variant.ref_aa | upper }}>{{variant.alt_aa | upper }}</td>
                        <td>{{ variant.depth | fmt_number }}</td>
                        <td>{{ variant.phenotypes | get_resistance_profile(level="name") }}</td>
                    </tr>
                {% endfor %}
            </tbody>
        </table>
    {% endif %}
</div>
{% endmacro %}


{% macro amrfinder_amr_prediction(result, extended=False) %}
<div>
    <h5 class="card-title">Genes</h5>
    {% if result.genes | length == 0 %}
        <p class="fw-light fst-italic">No resistance yeilding genes identified</p>
    {% else %}
    <table class="table table-striped fs-6">
        <thead>
            <td scope="col">Gene</td>
            <td scope="col">Method</td>
            {% if extended %}
                <td>Contig</td>
                <td>Start/ End</td>
            {% endif  %}
            <td scope="col">Identity</td>
            <td scope="col">Coverage</td>
            <td scope="col">Antibiotics</td>
            {% if extended %}
                <td>Name</td>
            {% endif  %}
        </thead>
        <tbody class="table-group-divider">
        {% for gene in result.genes %}
            <tr>
                <td>{{ accession_link(gene.accession, name=gene.gene_symbol) }}</td>
                <td>
                    {{ gene.method }}
                    <i class="bi bi-question-circle"
                        data-bs-toggle="tooltip"
                        data-bs-placement="top"
                        data-bs-html="true"
                        data-bs-title="{{ amrfinder_method_desc(method=gene.method) }}"></i>
                </td>
                {% if extended %}
                    <td class="tbl-border-left">{{ gene.contig_id }}</td>
                    <td class="tbl-border-right">{{ gene.ass_start_pos }}/ {{ gene.ass_end_pos }}</td>
                {% endif  %}
                <td>{{ gene.coverage }}%</td>
                <td>{{ gene.identity }}%</td>
                <td>{{ gene.phenotypes | get_resistance_profile(level="name") }}</td>
                {% if extended %}
                <td>{{ gene.close_seq_name }}</td>
                {% endif  %}
            </tr>
        {% endfor %}
        </tbody>
    </table>
    {% endif %}
</div>
{% endmacro %}


{% macro amrfinder_stress_prediction(result, extended=False) %}
<div>
    <h5 class="card-title">Genes</h5>
    {% if result.genes | length == 0 %}
        <p class="fw-light fst-italic">No resistance yeilding genes identified</p>
    {% else %}
    <table class="table table-striped">
        <thead>
            <td scope="col">Gene</td>
            <td scope="col">Type</td>
            <td scope="col">Method</td>
            <td scope="col">Identity</td>
            <td scope="col">Coverage</td>
            <td scope="col">Name</td>
        </thead>
        <tbody class="table-group-divider">
        {% for gene in result.genes %}
            <tr>
                <td>{{ accession_link(gene.accession, name=gene.gene_symbol) }}</td>
                <td>{{ gene.element_subtype }}</td>
                <td>
                    {{ gene.method }}
                    <i class="bi bi-question-circle"
                        data-bs-toggle="tooltip"
                        data-bs-placement="top"
                        data-bs-html="true"
                        data-bs-title="{{ amrfinder_method_desc(method=gene.method) }}"></i>
                </td>
                <td>{{ gene.coverage }}%</td>
                <td>{{ gene.identity }}%</td>
                <td>{{ gene.sequence_name | capitalize }}</td>
            </tr>
        {% endfor %}
        </tbody>
    </table>
    {% endif %}
</div>
{% endmacro %}


{% macro amrfinder_method_desc(method) %}
    {% if "BLAST" in method %}
        <p>The hit was identified in a protein (P) or nucleotide translated (X) or Nucleotide (BLASTN).  BLAST hits are hits that are &lt; 100% identical to a database protein, but at coverage &gt; 90%. The percent identity cutoff is, by default, 90%, but may be higher or lower if it was manually curated. Manual curation of BLAST parameters usually occurrs for <i>plus</i> genes where no HMM and cutoff have been curated.</p>
    {% elif "EXACT" in method %}
        <p>These mean that we have an amino-acid sequence identical protein in the AMRFinderPlus database.</p>
    {% elif "PARTIAL" in method %}
        <p>Hits &lt; 90% of the length of the database protein are called either PARTIAL if the hit is internal to a contig or PARTIAL_CONTIG_END if the gene could have been broken by a contig boundary.</p>
    {% elif "INTERNAL_STOP" in method %}
        <p>These are genes that when translated from genomic sequence have a stop codon before the end of the database protein.</p>
    {% elif method == "HMM" %}
        <p>HMM-only hits happen when the criteria for a BLAST hit is not met and an HMM match is above the curated cutoff for an HMM that has been created for that gene or gene family.</p>
    {% else %}
        <p>Test {{method}}</p>
    {% endif %}
{% endmacro %}


{% macro mykrobe_amr_prediction(result, extended=False) %}
<div>
    <h5 class="card-title">Genes</h5>
    {% if result.genes | length == 0 %}
        <p class="fw-light fst-italic">No resistance yeilding genes identified</p>
    {% else %}
    <table class="table table-striped fs-6">
        <thead>
            <td scope="col">Gene</td>
            <td scope="col">Method</td>
            {% if extended %}
                <td>Contig</td>
                <td>Start/ End</td>
            {% endif  %}
            <td scope="col">Identity</td>
            <td scope="col">Coverage</td>
            <td scope="col">Antibiotics</td>
            {% if extended %}
                <td>Name</td>
            {% endif  %}
        </thead>
        <tbody class="table-group-divider">
        {% for gene in result.genes %}
            <tr>
                <td>{{ accession_link(gene.accession, name=gene.gene_symbol) }}</td>
                <td>
                    {{ gene.method }}
                    <i class="bi bi-question-circle"
                        data-bs-toggle="tooltip"
                        data-bs-placement="top"
                        data-bs-html="true"
                        data-bs-title="{{ amrfinder_method_desc(method=gene.method) }}"></i>
                </td>
                {% if extended %}
                    <td class="tbl-border-left">{{ gene.contig_id }}</td>
                    <td class="tbl-border-right">{{ gene.ass_start_pos }}/ {{ gene.ass_end_pos }}</td>
                {% endif  %}
                <td>{{ gene.coverage }}%</td>
                <td>{{ gene.identity }}%</td>
                <td>{{ gene.phenotypes | get_resistance_profile(level="name") }}</td>
                {% if extended %}
                <td>{{ gene.close_seq_name }}</td>
                {% endif  %}
            </tr>
        {% endfor %}
        </tbody>
    </table>
    {% endif %}
    {# resistance variants #}
    <h5 class="card-title">Variants</h5>
    {% if result.variants | length == 0 %}
        <p class="fw-light fst-italic">No resistance yeilding mutations identified</p>
    {% else %}
        <table class="table table-striped">
            <thead>
                <tr>
                    <td scope="col">Variant</td>
                    <td scope="col">Type</td>
                    <td scope="col">Gene</td>
                    <td scope="col">Protein change</td>
                    <td scope="col">Method</td>
                    <td scope="col">KMER Depth<br/>(Alt / Ref)</td>
                    <td scope="col">Confidence</td>
                    <td scope="col">Antibiotics</td>
                </tr>
            </thead>
            <tbody class="table-group-divider">
                {% for variant in result.variants %}
                    {% set alt_depth=(variant.frequency * variant.depth) %}
                    {% set ref_depth=(variant.depth * (1 - variant.frequency)) %}
                    <tr>
                        <td>{{ variant.ref_nt | upper }}>{{variant.alt_nt | upper }}</td>
                        <td>{{ variant.variant_type[:3] }}</td>
                        <td>{{ accession_link(variant.close_seq_name, name=variant.gene_symbol) }}</td>
                        {% if variant.variant_type == 'substitution' %}
                            <td>{{ variant.ref_aa | upper }}>{{variant.alt_aa | upper }}</td>
                        {% else %}
                            <td>-</td>
                        {% endif %}
                        <td>{{ variant.method }}</td>
                        <td>{{ alt_depth | int | fmt_number  }} / {{ ref_depth | int | fmt_number }}</td>
                        <td>{{ variant.confidence }}</td>
                        <td>{{ variant.phenotypes[0].name | capitalize }}</td>
                    </tr>
                {% endfor %}
            </tbody>
        </table>
    {% endif %}
</div>
{% endmacro %}


{% macro tbprofiler_amr_prediction(result, extended=False) %}
<div>
    {# resistance variants #}
    <h5 class="card-title">Variants</h5>
    {% if result.variants | n_results_with_resistance == 0 and not extended %}
        <p class="fw-light fst-italic">No resistance yeilding mutations identified</p>
    {% else %}
        <table class="table table-striped">
            <thead>
                <tr>
                    {% if extended %}
                        <td scope="col"></td>
                    {% endif %}
                    <td scope="col">Tags</td>
                    <td scope="col">Variant</td>
                    <td scope="col">Protein change</td>
                    <td scope="col">Gene</td>
                    <td scope="col">Read depth</td>
                    <td scope="col">Freq</td>
                    <td scope="col">Passed QC</td>
                    <td scope="col">Antibiotics</td>
                    {% if extended %}
                        <td scope="col"></td>
                    {% endif %}
                </tr>
            </thead>
            <tbody class="table-group-divider">
                {% for variant in result.variants %}
                    {% if variant.phenotypes | length > 0 or extended %}
                    {% if variant.verified == "passed" %}
                        {% set row_class = "table-success" %}
                    {% elif variant.verified == "failed" %}
                        {% set row_class = "table-danger" %}
                    {% endif %}
                    <tr class="{{ row_class }}">
                        {% if extended %}
                            <td>
                                <input type="checkbox" class="form-check-input col-auto" name="tbprofiler-{{ variant.id }}" id="tbprofiler-{{ variant.id }}" onclick="selectVariant(this)">
                            </td>
                        {% endif %}
                        <td>
                            {% if variant.verified == "failed" and variant.reason %}
                                <span class="ms-2 badge text-bg-warning"
                                        data-bs-toggle="tooltip" data-bs-placement="top"
                                        data-bs-title="{{ variant.reason | capitalize }}" >
                                    {{ variant.reason[0] | upper }}
                                </span>
                            {% endif %}
                            {# Add WHO class as badges #}
                            {% for phenotype in variant.phenotypes %}
                                {% if phenotype.note and phenotype.note != "" %}
                                    <span class="badge text-bg-secondary"
                                        data-bs-toggle="tooltip" data-bs-placement="top"
                                        data-bs-title="{{ phenotype.note }}">
                                        {{ phenotype.note | lookup_who_group }}
                                    </span>
                                {% endif %}
                            {% endfor %}
                        </td>
                        <td>
                            {{ variant.hgvs_nt_change }}
                        </td>
                        {% if variant.hgvs_aa_change %}
                            <td>{{ variant.hgvs_aa_change }}</td>
                        {% else %}
                            <td>-</td>
                        {% endif %}
                        <td>{{ accession_link(variant.close_seq_name, name=variant.reference_sequence) }}</td>
                        <td>{{ variant.depth | int }}</td>
                        <td>{{ (variant.frequency * 100) | round(1) }}</td>
                        {% if variant.passed_qc  %}
                            <td><i class="bi bi-check text-success"></i></td>
                        {% else  %}
                            <td><i class="bi bi-x text-danger"></i></td>
                        {% endif  %}
                        <td>
                            <ul class="no-bullets">
                            {% for phenotype in variant.phenotypes %}
                                <li>
                                <span data-bs-toggle="tooltip" data-bs-placement="top" 
                                      data-bs-title="{{ phenotype.group }}">
                                    {{ phenotype.name | capitalize }}
                                </span>
                                {# Add tag if antibiotic is user annotated #}
                                {% if phenotype.annotation_type == "user" %}
                                    <span class="badge text-bg-primary"
                                          data-bs-toggle="tooltip" data-bs-placement="top"
                                          data-bs-title="{{ phenotype.annotation_author }}">
                                        <i class="bi bi-person-fill"></i>
                                    </span>
                                {% endif %}
                                </li>
                            {% endfor %}
                            </ul>
                        </td>
                        {% if extended and display_igv %}
                            <td>
                                {% set variant_id = "tbprofiler-{}".format(variant["id"]) %}
                                <a href="{{ url_for('alignviewers.igv', sample_id=sample.sample_id, variant_id=variant_id) }}"
                                   target="_blank" class="btn btn-sm btn-secondary" type="button">
                                   IGV
                                </a>
                            </td>
                        {% endif %}
                    </tr>
                    {% endif %}
                {% endfor %}
            </tbody>
        </table>
    {% endif %}
</div>
{% endmacro %}


{% macro stx_card(pred) %}
<div class="card col-md-auto">
    <div class="card-header">
        {{ pred.type | upper }} typing [<i>{{ pred.software }}</i>]
    </div>
    <div class="card-body">
        <table class="table table-sm">
            <thead>
                <td>Gene</td>
                <td>Coverage</td>
                <td>Identity</td>
                <td>Name</td>
            </thead>
            <tbody class="table-group-divider">
                {% set gene=pred.result %}
                <tr>
                    <td>{{ accession_link(gene.accession, name=gene.gene_symbol) }}</td>
                    <td>{{ gene.coverage }}%</td>
                    <td>{{ gene.identity }}%</td>
                    <td>{{ gene.sequence_name | capitalize }}</td>
                </tr>
            </tbody>
        </table>
    </div>
</div>
{% endmacro %}


{% macro analysis_meta_pipeline(sample_info) %}
{% set meta=sample_info["run_metadata"] %}
<div class="card">
    <div class="card-body">
        <h6 class="card-title">
            <i class="bi bi-server me-2"></i>
            <a class="br-card-link" href="https://github.com/genomic-medicine-sweden/JASEN" target="_blank">
                Pipeline
            </a>
        </h5>
        <table class="table">
            <tbody>
                <tr>
                    <th>Pipeline</th>
                    <td>{{ meta.run.pipeline }}</td>
                </tr>
                <tr>
                    <th>Version</th>
                    <td>{{ meta.run.version }}</td>
                </tr>
                <tr>
                    <th>Profile</th>
                    <td>{{ meta.run.analysis_profile }}</td>
                </tr>
            </tbody>
        </table>
    </div>
</div>
{% endmacro %}

{% macro analysis_meta_databases(sample_info) %}
{% set meta=sample_info["run_metadata"]%}
<div class="card">
    <div class="card-body">
        <h6 class="card-title">
            <i class="bi bi-server me-2"></i>Databases
        </h6>
        <table class="table">
            <tbody>
                {% for db in meta["databases"] %}
                <tr>
                    <th>{{ db["name"] }}</th>
                    <td>{{ make_db_link(db) }}</td>
                </tr>
                {% endfor %}
            </tbody>
        </table>
    </div>
</div>
{% endmacro %}

{% macro mlst_card(mlst) %}
<div class="card col-md-auto">
    <div class="card-header">MLST</div>
    <div class="card-body">
        <div class="p-2">
            <strong>ST</strong>
            {{ mlst.sequence_type }}
        </div>
        <div class="p-2">
            <table class="table table-sm">
                <thead>
                    <th>Gene</th>
                    {% for name in mlst.alleles %}
                        <td scope="col">{{name}}</td>
                    {% endfor %}
                </thead>
                <tbody class="table-group-divider">
                    <th>Allele No</th>
                    {% for value in mlst.alleles.values() %}
                        {% if value is list %}
                            <td class="fst-italic">{{ value | join("/") }}</td>
                        {% else %}
                            <td>{{ value }}</td>
                        {% endif %}
                    {% endfor %}
                </tbody>
            </table>
        </div>
    </div>
</div>
{% endmacro %}


{% macro similar_samples_card(data) %}
<div id="similar-samples-card"
     class="card col-md-auto {% if data is none %}border-warning{% endif %}" 
     style="width: 24rem;">
    <div class="card-header">
        Similar samples {% if data is not none %}[<i>{{ data.typing_method }}</i>]{% endif %}
    </div>
    <div id="similar-samples-cluster" class="card-body" data-job-id="{{data.job.id}}">
        <div class="d-flex justify-content-center loading my-5">
            <div class="spinner-border text-success" role="status">
                <span class="visually-hidden">Loading...</span>
            </div>
        </div>
        <div class="content">
            <div id="tree-body" class="job-success">
            </div>
        </div>
    </div>
</div>
{% endmacro %}


{% macro qc_card_postalign(qc_info) %}
<div class="card">
    <div class="card-header">PostQc Alignment</div>
    <div class="card-body">
        {% set pct_above_hist=qc_info.result.pct_above_x %}
        {# create entries for tabular data #}
        <table class="table table-sm table-striped">
            <thead>
                <tr>
                    <th scope="col">Metric</th>
                    <th scope="col">Value</th>
                </tr>
            </thead>
            <tbody>
            {% for name, value in qc_info.result.items() %}
                {# skip data that should be plotted #}
                {% if name != "pct_above_x" %}
                    <tr>
                        <th scope="row">{{ name | camelcase_to_text | capitalize }}</th>
                        <td>{{ value | fmt_number(sig_digits=2) }}</td>
                    </tr>
                {% endif %}
            {% endfor %}
            </tbody>
        </table>
        <div id="qcPlot"></div>
        <script>
            //const qcPlot = document.getElementById("qc-plot")
            const rawData = {{pct_above_hist | tojson }}
            const data = {
                type: 'bar',
                x: Object.keys(rawData).map(k => `${k}X`), 
                y: Object.values(rawData).map(Number), 
                text: Object.values(rawData).map(num  => `${Math.round(Number(num), 1)}%`),
                textposition: 'auto',
                marker: {
                    color: "#c0ddd1",
                    line: { width: 4}
                }
            };
            const layout = { 
                title: 'Genome coverage', 
                autosize: false,
                width: 400,
                height: 200,
                margin: {
                    l: 10,
                    r: 10,
                    b: 30,
                    t: 30,
                    pad: 4
                },
                hoverinfo: '',
                yaxis: {
                    title: "Percent bases",
                    titlefont: { size: 12 },
                    automargin: true
                },
                xaxis: {
                    title: "Coverage",
                    titlefont: { size: 12 },
                    automargin: true
                },
                font: { size: 12 }}
            Plotly.newPlot('qcPlot', [data], layout, { staticPlot: true} )
        </script>
    </div>
</div>
{% endmacro %}


{% macro qc_card_generic(qc_info) %}
<div class="card">
    <div class="card-header">{{qc_info.software | capitalize}}</div>
    <div class="card-body">
        <table class="table table-sm table-striped">
            <thead>
                <tr>
                    <th scope="col">Metric</th>
                    <th scope="col">Value</th>
                </tr>
            </thead>
            <tbody>
            {% for name, value in qc_info.result.items() %}
            <tr>
                <th scope="row">{{ name | camelcase_to_text | capitalize }}</th>
                <td>{{ value | fmt_number(sig_digits=2) }}</td>
            </tr>
            {% endfor %}
            </tbody>
        </table>
    </div>
</div>
{% endmacro %}


{% macro cgmlst_qc_card(result) %}
<div class="card">
    <div class="card-header">
        {{ result.type }} QC
    </div>
    <div class="card-body">
        <table class="table table-sm table-striped">
            <tr>
                <th scope="row">N Called Alleles</th>
                <td>{{ result.result.alleles | cgmlst_count_called }}</td>
            </tr>
            <tr>
                <th scope="row">N Missing Alleles</th>
                <td>{{ result.result.alleles | cgmlst_count_missing - result.result.n_novel }}</td>
            </tr>
        </table>
    </div>
</div>
{% endmacro %}