{% macro accession_link(accession, name=None) %}
    <a class="text-dark" target="_blank" href="https://www.ncbi.nlm.nih.gov/protein/{{accession}}">
        {% if name == None %}
            {{ accession }}
        {% else %}
            {{ name }}
        {% endif %}
    </a>
{% endmacro %}

{% macro resistance_summary_row(gene)%}
    <td>{{ accession_link(gene.accession) }}</td>
    <td>{{ gene.depth | fmt_null_values }}</td>
    <td>{{ gene.identity | fmt_number }}%</td>
    <td>{{ gene.coverage | fmt_number }}%</td>
    <td class="fst-italic fw-lighter">{{ gene.software }}</td>
{% endmacro %}

{% macro resistance_summary_card(resistance, summary=None) %}
<div class="card">
    <div class="card-header">AMR profile summary</div>
    <div class="card-body">
        <div class="table-responsive">
            <table class="table table-sm table-hover">
                <thead>
                    {% for antib_class in summary %}
                        <th>{{ antib_class | capitalize }}</th>
                    {% endfor %}
                </thead>
                <tbody>
                    {% for antibs in zip_longest(*summary.values()) %}
                        <tr>
                            {% for antib in antibs %}
                            <td>
                                {% if antib %}
                                    {{ antib.gene_symbol }}
                                    <span class="badge rounded-pill bg-secondary">
                                        {{antib.software | length}}
                                    </span>
                                {% endif %}
                            </td>
                            {% endfor %}
                        </tr>
                    {% endfor %}
                </tbody>
            </table>
        </div>
    </div>
</div>
{% endmacro %}


{% macro resistance_table_card(resistance, summary=None) %}
<div class="card">
    {% set res=resistance.result%}
    <div class="card-header">{{ resistance.type | camelcase_to_text | capitalize }}</div>
    <div class="card-body">
        <h5 class="card-title">Predicted genes</h5>
        <table class="table">
            <thead>
                <tr>
                    <td scope="col">Gene</td>
                    <td scope="col">Name</td>
                    <td scope="col">Class</td>
                    <td scope="col">Accession</td>
                    <td scope="col">Identity</td>
                    <td scope="col">Coverage</td>
                </tr>
            </thead>
            <tbody>
            {% for gene in res.genes %}
            <tr>
                <td>{{ gene.gene_symbol }}</td>
                <td>{{ gene.sequence_name }}</td>
                <td>
                    {% if gene.res_class == gene.res_subclass %}
                        {{ gene.res_subclass | lower | capitalize }}
                    {% else %}
                        {{ gene.res_class | lower | capitalize }} / {{ gene.res_subclass | lower | capitalize }} 
                    {% endif %}
                </td>
                <td>{{ accession_link(gene.accession) }}</td>
                <td>{{ gene.coverage }}</td>
                <td>{{ gene.identity }}</td>
            </tr>
            {% endfor %}
            </tbody>
        </table>
        <h5 class="card-title">Predicted mutations</h5>
            {% if resistance.mutations | length == 0 %}
                <p class="fw-light fst-italic">No resistance yeilding mutations identified</p>
            {% else %}
                <table class="table">
                    <thead>
                        <tr>
                            <td scope="col">Mutation</td>
                            <td scope="col">Change</td>
                            <td scope="col">Type</td>
                            <td scope="col">Gene</td>
                            <td scope="col">Depth</td>
                        </tr>
                    </thead>
                    <tbody>
                    {% for mut in res.mutations %}
                    <tr>
                        <td>{{ mut.refCodon }}<b>{{ mut.position }}</b>{{ mut.altCodon }}
                            <span class="badge bg-danger" data-bs-toggle="tooltip" data-bs-placement="top" 
                            >
                                R {{ mut.phenotypes | length }}
                            </span>
                        </td>
                        <td>{{ mut.ref_codon | nt_to_aa }}<i class="bi bi-arrow-right mx-2"></i>{{ mut.altCodon | nt_to_aa }}</td>
                        <td>{{ mut.variant_type }}</td>
                        <td>{{ ", ".join(mut.genes) }}</td>
                        <td>{{ mut.depth }}</td>
                    </tr>
                    {% endfor %}
                    </tbody>
                </table>
            {% endif %}
    </div>
</div>
{% endmacro %}


{% macro virulence_card(pred) %}
<div class="card">
    <div class="card-header">{{ pred.software | capitalize }} [ <i>{{pred.type}}</i> ]</div>
    <div class="card-body">
    {% if pred.software == "virulencefinder" %}
        {{ virulencefinder_prediction(pred.result) }}
    {% elif pred.software == "amrfinder" and pred.type == "VIRULENCE" %}
        {{ amrfinder_virulence_prediction(pred.result) }}
    {% endif %}
    </div>
</div>
{% endmacro %}


{% macro amrfinder_virulence_prediction(res, extended=False) %}
<table class="table table-sm table-striped">
    <thead>
        <tr>
            <td scope="col">Gene</td>
            <td scope="col">Method</td>
            {% if extended %}
                <td>Contig</td>
                <td>Start/ End</td>
            {% endif  %}
            <td scope="col">Identity</td>
            <td scope="col">Coverage</td>
            <td scope="col">Name</td>
        </tr>
    </thead>
    <tbody class="table-group-divider">
    {% for gene in res.genes %}
    <tr>
        <td>{{ accession_link(gene.accession, name=gene.gene_symbol) }}</td>
        <td>
            {{ gene.method }}
            <i class="bi bi-question-circle"
                data-bs-toggle="tooltip"
                data-bs-placement="top"
                data-bs-html="true"
                data-bs-title="{{ amrfinder_method_desc(method=gene.method) }}"></i>
        </td>
        {% if extended %}
            <td class="tbl-border-left">{{ gene.contig_id }}</td>
            <td class="tbl-border-right">{{ gene.ass_start_pos }}/ {{ gene.ass_end_pos }}</td>
        {% endif  %}
        <td>{{ gene.coverage }}%</td>
        <td>{{ gene.identity }}%</td>
        <td>{{ gene.sequence_name  | capitalize }}</td>
    </tr>
    {% endfor %}
    </tbody>
</table>
{% endmacro %}


{% macro virulencefinder_prediction(res, extended=False) %}
<table class="table table-sm table-striped">
    <thead>
        <tr>
            <td scope="col">Gene</td>
            {% if extended %}
                <td>Contig</td>
                <td>Start/ End</td>
            {% endif  %}
            <td scope="col">Identity</td>
            <td scope="col">Coverage</td>
            <td scope="col">Function</td>
        </tr>
    </thead>
    <tbody class="table-group-divider">
    {% for gene in res.genes %}
    <tr>
        <td>{{ accession_link(gene.accession, name=gene.gene_symbol) }}</td>
        {% if extended %}
            <td class="tbl-border-left">{{ gene.contig_id }}</td>
            <td class="tbl-border-right">{{ gene.ass_start_pos }}/ {{ gene.ass_end_pos }}</td>
        {% endif  %}
        <td>{{ gene.coverage }}%</td>
        <td>{{ gene.identity }}%</td>
        <td>{{ gene.sequence_name | capitalize }}</td>
    </tr>
    {% endfor %}
    </tbody>
</table>
{% endmacro %}


{% macro species_prediction_card(predictions) %}
<div class="card">
    <div class="card-header">Species prediction</div>
    <div class="card-body">
        <table class="table table-sm">
            <thead>
                <th>Specie</th>
                <th>Fraction of reads</th>
                <th>N reads</th>
            </thead>
            <tbody>
                {% for pred in predictions %}
                    <tr class="{% if predictions | length > 1 and loop.index == 1 %}table-success{% endif %}">
                        <td> <a class="text-decoration-underline text-dark fst-italic" 
                               href="https://www.ncbi.nlm.nih.gov/datasets/taxonomy/{{pred.taxonomy_id}}/" 
                               target="_blank">{{pred.scientific_name}}</a></td>
                        <td>{{ (pred.fraction_total_reads * 100) | fmt_number(2) }}%</td>
                        <td>{{ pred.kraken_assigned_reads | fmt_to_human_readable }}</td>
                    </tr>
                {% endfor %}
            </tbody>
        </table>
    </div>
</div>
{% endmacro %}


{% macro amr_prediction_card(pred) %}
<div class="card">
    <div class="card-header">{{ pred.software | capitalize }} [ <i>{{pred.type}}</i> ]</div>
    <div class="card-body">
    {% if pred.software == "resfinder" %}
        {{ resfinder_prediction(pred.result) }}
    {% elif pred.software == "amrfinder" and pred.type == "AMR" %}
        {{ amrfinder_amr_prediction(pred.result) }}
    {% elif pred.software == "amrfinder" and pred.type == "STRESS" %}
        {{ amrfinder_stress_prediction(pred.result) }}
    {% endif %}
    </div>
</div>
{% endmacro %}


{% macro resfinder_prediction(result) %}
<div>
    <h5 class="card-title">Genes</h5>
    <table class="table table-sm table-striped">
        <caption>Database: {{result.genes[0].ref_database}}</caption>
        <thead>
            <td scope="col">Gene</td>
            <td scope="col">Accession</td>
            <td scope="col">Identity</td>
            <td scope="col">Coverage</td>
            <td scope="col">Depth</td>
            <td scope="col">Antibiotics</td>
        </thead>
        <tbody class="table-group-divider">
        {% for gene in result.genes %}
            <tr>
                <td>{{ gene.gene_symbol }}</td>
                <td>{{ accession_link(gene.accession) }}</td>
                <td>{{ gene.coverage }}%</td>
                <td>{{ gene.identity }}%</td>
                <td>{{ gene.depth | fmt_number }}</td>
                <td>{{ gene.phenotypes | get_resistance_profile(level="name") }}</td>
            </tr>
        {% endfor %}
        </tbody>
    </table>
    <h5 class="card-title">Variants</h5>
    {% if result.mutations | length == 0 %}
        <p class="fw-light fst-italic">No resistance yeilding mutations identified</p>
    {% else %}
        <table class="table table-striped">
            <caption>Database: {{ result.mutations[0].ref_database }}</caption>
            <thead>
                <tr>
                    <td scope="col">Variant</td>
                    <td scope="col">Gene</td>
                    <td scope="col">Protein change</td>
                    <td scope="col">Depth</td>
                    <td scope="col">Antibiotics</td>
                </tr>
            </thead>
            <tbody class="table-group-divider">
                {% for variant in result.mutations %}
                    <tr>
                        <td>{{ variant.name }}</td>
                        <td>{{ accession_link(variant.close_seq_name, name=variant.gene_symbol) }}</td>
                        <td>{{ variant.ref_aa | upper }}>{{variant.alt_aa | upper }}</td>
                        <td>{{ variant.depth | fmt_number }}</td>
                        <td>{{ variant.phenotypes | get_resistance_profile(level="name") }}</td>
                    </tr>
                {% endfor %}
            </tbody>
        </table>
    {% endif %}
</div>
{% endmacro %}


{% macro amrfinder_amr_prediction(result, extended=False) %}
<div>
    <h5 class="card-title">Genes</h5>
    {% if result.genes | length == 0 %}
        <p class="fw-light fst-italic">No resistance yeilding genes identified</p>
    {% else %}
    <table class="table table-sm table-striped">
        <thead>
            <td scope="col">Gene</td>
            <td scope="col">Method</td>
            {% if extended %}
                <td>Contig</td>
                <td>Start/ End</td>
            {% endif  %}
            <td scope="col">Identity</td>
            <td scope="col">Coverage</td>
            <td scope="col">Antibiotics</td>
            {% if extended %}
                <td>Name</td>
            {% endif  %}
        </thead>
        <tbody class="table-group-divider">
        {% for gene in result.genes %}
            <tr>
                <td>{{ accession_link(gene.accession, name=gene.gene_symbol) }}</td>
                <td>
                    {{ gene.method }}
                    <i class="bi bi-question-circle"
                        data-bs-toggle="tooltip"
                        data-bs-placement="top"
                        data-bs-html="true"
                        data-bs-title="{{ amrfinder_method_desc(method=gene.method) }}"></i>
                </td>
                {% if extended %}
                    <td class="tbl-border-left">{{ gene.contig_id }}</td>
                    <td class="tbl-border-right">{{ gene.ass_start_pos }}/ {{ gene.ass_end_pos }}</td>
                {% endif  %}
                <td>{{ gene.coverage }}%</td>
                <td>{{ gene.identity }}%</td>
                <td>{{ gene.phenotypes | get_resistance_profile(level="name") }}</td>
                {% if extended %}
                <td>{{ gene.close_seq_name }}</td>
                {% endif  %}
            </tr>
        {% endfor %}
        </tbody>
    </table>
    {% endif %}
</div>
{% endmacro %}


{% macro amrfinder_stress_prediction(result) %}
<div>
    <h5 class="card-title">Genes</h5>
    {% if result.genes | length == 0 %}
        <p class="fw-light fst-italic">No resistance yeilding genes identified</p>
    {% else %}
    <table class="table table-sm table-striped">
        <thead>
            <td scope="col">Gene</td>
            <td scope="col">Type</td>
            <td scope="col">Method</td>
            <td scope="col">Identity</td>
            <td scope="col">Coverage</td>
            <td scope="col">Name</td>
        </thead>
        <tbody class="table-group-divider">
        {% for gene in result.genes %}
            <tr>
                <td>{{ accession_link(gene.accession, name=gene.gene_symbol) }}</td>
                <td>{{ gene.element_subtype }}</td>
                <td>
                    {{ gene.method }}
                    <i class="bi bi-question-circle"
                        data-bs-toggle="tooltip"
                        data-bs-placement="top"
                        data-bs-html="true"
                        data-bs-title="{{ amrfinder_method_desc(method=gene.method) }}"></i>
                </td>
                <td>{{ gene.coverage }}%</td>
                <td>{{ gene.identity }}%</td>
                <td>{{ gene.sequence_name | capitalize }}</td>
            </tr>
        {% endfor %}
        </tbody>
    </table>
    {% endif %}
</div>
{% endmacro %}


{% macro amrfinder_method_desc(method) %}
    {% if "BLAST" in method %}
        <p>The hit was identified in a protein (P) or nucleotide translated (X) or Nucleotide (BLASTN).  BLAST hits are hits that are &lt; 100% identical to a database protein, but at coverage &gt; 90%. The percent identity cutoff is, by default, 90%, but may be higher or lower if it was manually curated. Manual curation of BLAST parameters usually occurrs for <i>plus</i> genes where no HMM and cutoff have been curated.</p>
    {% elif "EXACT" in method %}
        <p>These mean that we have an amino-acid sequence identical protein in the AMRFinderPlus database.</p>
    {% elif "PARTIAL" in method %}
        <p>Hits &lt; 90% of the length of the database protein are called either PARTIAL if the hit is internal to a contig or PARTIAL_CONTIG_END if the gene could have been broken by a contig boundary.</p>
    {% elif "INTERNAL_STOP" in method %}
        <p>These are genes that when translated from genomic sequence have a stop codon before the end of the database protein.</p>
    {% elif method == "HMM" %}
        <p>HMM-only hits happen when the criteria for a BLAST hit is not met and an HMM match is above the curated cutoff for an HMM that has been created for that gene or gene family.</p>
    {% else %}
        <p>Test {{method}}</p>
    {% endif %}
{% endmacro %}